src/compiler/tree.cc

```diff
@@ -30,23 +31,62 @@ using namespace ir;
 
 typedef Selector<CallShape> CallSelector;
 
-class GrowerVisitor : public TraversingVisitor {
+class GrowerVisitor : protected TraversingVisitor {
  public:
   explicit GrowerVisitor(Method* as_check_failure)
       : as_check_failure_(as_check_failure) {}
 
   Set<Class*> found_classes() const { return found_classes_; }
   Set<Method*> found_methods() const { return found_methods_; }
+  // Selectors that only are valid for the given set of methods.
+  // The optimizer changed the dynamic call to a static call, which means that
+  // it knew that the selector goes to that function. We still need to
+  // verify that the holder was instantiated.
+  Map<CallSelector, UnorderedSet<Method*>> found_filtered_selectors() const {
+    return found_filtered_selectors_;
+  }
   Set<CallSelector> found_selectors() const { return found_selectors_; }
 
+  void grow(Method* method) {
+    current_method_ = method;
+    visit(method);
+  }
+
+ protected:
   void visit_CallConstructor(CallConstructor* node) {
     found_classes_.insert(node->klass());
     found_methods_.insert(node->target()->target());
     TraversingVisitor::visit_CallConstructor(node);
   }
 
+  bool is_super_call(CallStatic* node) const {
+    auto current_holder = current_method_->holder();
+    auto target = node->target()->target();
+    if (!is_This(node->arguments()[0], current_holder, target)) return false;
+    // Make sure this is actually a super call (and not a sub call).
+    auto current_class = current_holder;
```

Comment:
This would probably feel natural to have on ir::Class.

Reply:
done.